File: web/src/audio/mixer.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

/**
 * What the monitor desk amounts to in amplitudes.
 *
 * The worklet mixes with these, the editor draws with them and the panel reads them back, so
 * what is heard, what is drawn and what the faders say cannot disagree.
 */

import { MAX_GAIN_DB, MIN_GAIN_DB } from '../core/types.js';
import type { ClipId, ClipStrips, MixerSettings, MixerStrip, ReferenceId } from '../core/types.js';

/** Which of a clip's two strips: the take as edited or as sung. */
export type VocalStrip = 'processed' | 'original';

/** What each vocal strip is called. */
export const VOCAL_NAMES: Readonly<Record<VocalStrip, string>> = {
  processed: 'Processed',
  original: 'Original',
};

/**
 * What the click strip is called.
 *
 * @remarks The name the transport control already uses, because two names for one sound is a
 * thing to work out rather than a thing to read.
 */
export const CLICK_NAME = 'Metronome';

/** What one strip contributes to each output channel. */
export interface StripLevel {
  left: number;
  right: number;
  /** Whether the strip is heard at all, which is what mute, solo and a closed fader decide. */
  audible: boolean;
}

/** What one clip's two strips contribute. */
export interface ClipLevels {
  processed: StripLevel;
  original: StripLevel;
}

/**
 * Amplitudes every strip contributes, mute and solo already resolved.
 *
 * @remarks Read on the audio thread, so looking a source up allocates nothing: a source with no
 * entry on the desk reads the shared default.
 */
export interface MixLevels {
  clips: ReadonlyMap<ClipId, ClipLevels>;
  references: ReadonlyMap<ReferenceId, StripLevel>;
  click: StripLevel;
  /** What a clip with no entry on the desk contributes. */
  defaultClip: ClipLevels;
  /** What a reference with no entry on the desk contributes. */
  defaultReference: StripLevel;
}

/**
 * Linear amplitude of a level in decibels, with the floor reading as silence.
 *
 * @remarks Mirrors `axys_core::units::decibels_to_amplitude`, because the worklet mixes on the
 * audio thread and cannot call into the core for a multiplier.
 */
export function amplitude(decibels: number): number {
  if (!Number.isFinite(decibels)) return 1;
  const clamped = Math.min(Math.max(decibels, MIN_GAIN_DB), MAX_GAIN_DB);
  return clamped <= MIN_GAIN_DB ? 0 : 10 ** (clamped / 20);
}

/** Level the click is heard at until it is moved, mirroring the core's own default. */
export const DEFAULT_CLICK_DB = -11;

/** A strip at unity, centred, neither muted nor soloed. */
export const UNITY_STRIP: MixerStrip = { gainDb: 0, pan: 0, mute: false, solo: false };

/** The desk a project starts with, for the moments before one is open. */
export const DEFAULT_MIXER: MixerSettings = {
  clips: [],
  references: [],
  click: { gainDb: DEFAULT_CLICK_DB, pan: 0, mute: false, solo: false },
};

/**
 * A clip's track, or the one it starts with: the edited take up and the original muted.
 *
 * @remarks Mirrors `MixerSettings::clip` in the core.
 */
export function clipStrips(mixer: MixerSettings, clip: ClipId): ClipStrips {
  return (
    mixer.clips.find((entry) => entry.clip === clip) ?? {
      clip,
      processed: UNITY_STRIP,
      original: { ...UNITY_STRIP, mute: true },
    }
  );
}

/** A reference's strip, or the unity strip it starts with. */
export function referenceStrip(mixer: MixerSettings, reference: ReferenceId): MixerStrip {
  return mixer.references.find((entry) => entry.reference === reference)?.strip ?? UNITY_STRIP;
}

/** The desk with one clip strip replaced, adding the clip's entry when it had none. */
export function withClipStrip(
  mixer: MixerSettings,
  clip: ClipId,
  which: VocalStrip,
  strip: MixerStrip,
): MixerSettings {
  const track = { ...clipStrips(mixer, clip), [which]: strip };
  const clips = mixer.clips.some((entry) => entry.clip === clip)
    ? mixer.clips.map((entry) => (entry.clip === clip ? track : entry))
    : [...mixer.clips, track];
  return { ...mixer, clips };
}

/** The desk with one reference strip replaced, adding its entry when it had none. */
export function withReferenceStrip(
  mixer: MixerSettings,
  reference: ReferenceId,
  strip: MixerStrip,
): MixerSettings {
  const references = mixer.references.some((entry) => entry.reference === reference)
    ? mixer.references.map((entry) =>
        entry.reference === reference ? { reference, strip } : entry,
      )
    : [...mixer.references, { reference, strip }];
  return { ...mixer, references };
}

/** Whether any strip on the desk is soloed, which is what silences the ones that are not. */
export function soloed(mixer: MixerSettings): boolean {
  return (
    mixer.click.solo ||
    mixer.clips.some((entry) => entry.processed.solo || entry.original.solo) ||
    mixer.references.some((entry) => entry.strip.solo)
  );
}

/**
 * What each strip contributes to the two output channels.
 *
 * @remarks Pan is equal power, so a strip swept across the field holds its loudness rather than
 * dipping through the middle. Any solo silences every strip that is not soloed, which is what a
 * solo means everywhere else.
 */
export function mixLevels(mixer: MixerSettings): MixLevels {
  const solo = soloed(mixer);
  const clips = new Map<ClipId, ClipLevels>();
  for (const entry of mixer.clips) {
    clips.set(entry.clip, {
      processed: levelOf(entry.processed, solo),
      original: levelOf(entry.original, solo),
    });
  }
  const references = new Map<ReferenceId, StripLevel>();
  for (const entry of mixer.references) {
    references.set(entry.reference, levelOf(entry.strip, solo));
  }
  const fresh = clipStrips({ ...mixer, clips: [] }, -1);
  return {
    clips,
    references,
    click: levelOf(mixer.click, solo),
    defaultClip: {
      processed: levelOf(fresh.processed, solo),
      original: levelOf(fresh.original, solo),
    },
    defaultReference: levelOf(UNITY_STRIP, solo),
  };
}

/** What one clip contributes, reading the default for a clip the desk has no entry for. */
export function clipLevels(levels: MixLevels, clip: ClipId): ClipLevels {
  return levels.clips.get(clip) ?? levels.defaultClip;
}

/** What one reference contributes, reading the default for one the desk has no entry for. */
export function referenceLevel(levels: MixLevels, reference: ReferenceId): StripLevel {
  return levels.references.get(reference) ?? levels.defaultReference;
}

function levelOf(strip: MixerStrip, solo: boolean): StripLevel {
  const open = solo ? strip.solo : !strip.mute;
  const gain = open ? amplitude(strip.gainDb) : 0;
  const angle = ((Math.min(Math.max(strip.pan, -1), 1) + 1) * Math.PI) / 4;
  return { left: gain * Math.cos(angle), right: gain * Math.sin(angle), audible: gain > 0 };
}

/** Which of a clip's two strips is being heard, which is what the editor draws solid. */
export function vocalMonitor(
  mixer: MixerSettings,
  clip: ClipId,
): 'processed' | 'original' | 'both' | 'neither' {
  const levels = clipLevels(mixLevels(mixer), clip);
  if (levels.processed.audible && levels.original.audible) return 'both';
  if (levels.processed.audible) return 'processed';
  if (levels.original.audible) return 'original';
  return 'neither';
}

File: crates/axys-core/src/mixer.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

//! Monitor levels for everything the transport plays.
//!
//! The desk has a track per vocal clip carrying two strips, the take as edited and the take as
//! sung, a strip per reference, and one for the click. Each strip has the controls a desk has:
//! level, pan, mute and solo. The mixer is monitoring rather than an edit to the take, so it never
//! reaches the render plan and an export is unchanged by it. It lives in the project document
//! because how a take is listened to is part of the work, and it is set by an edit operation like
//! anything else, so it undoes.

use serde::{Deserialize, Serialize};

use crate::clip::{ClipId, ReferenceId, MAX_CLIPS, MAX_REFERENCES};
use crate::units::decibels_to_amplitude;
use crate::{limits, AxysError, Result};

/// Level the click is heard at until it is moved, in decibels.
pub const DEFAULT_CLICK_DB: f64 = -11.0;

/// One audio source on the monitor desk.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MixerStrip {
    /// Level in decibels; 0.0 is unity and [`limits::MIN_GAIN_DB`] is silence.
    pub gain_db: f64,
    /// Position across the stereo field, -1.0 hard left to 1.0 hard right.
    pub pan: f64,
    /// Silences the strip.
    pub mute: bool,
    /// Silences every strip that is not soloed.
    pub solo: bool,
}

impl MixerStrip {
    /// A strip at a level, muted or not, centred and unsoloed.
    pub fn new(gain_db: f64, mute: bool) -> Self {
        Self {
            gain_db,
            pan: 0.0,
            mute,
            solo: false,
        }
    }

    /// Linear amplitude of the strip's level.
    pub fn amplitude(&self) -> f64 {
        decibels_to_amplitude(self.gain_db)
    }

    fn validated(&self) -> Result<Self> {
        if !self.gain_db.is_finite() || !self.pan.is_finite() {
            return Err(AxysError::Invalid("mixer setting is not finite".into()));
        }
        Ok(Self {
            gain_db: self.gain_db.clamp(limits::MIN_GAIN_DB, limits::MAX_GAIN_DB),
            pan: self.pan.clamp(-1.0, 1.0),
            ..*self
        })
    }
}

impl Default for MixerStrip {
    fn default() -> Self {
        Self::new(0.0, false)
    }
}

/// A vocal clip's track on the desk: the take as edited and as sung, side by side.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClipStrips {
    /// The clip the track belongs to.
    pub clip: ClipId,
    /// The take as the edits make it sound.
    pub processed: MixerStrip,
    /// The take as it was sung, on the same transport clock.
    pub original: MixerStrip,
}

impl ClipStrips {
    /// The track a clip starts with: the edited take up, the original muted.
    pub fn new(clip: ClipId) -> Self {
        Self {
            clip,
            processed: MixerStrip::default(),
            // The edited take is what the editor is for, so the original starts muted and is
            // brought in to compare against it.
            original: MixerStrip::new(0.0, true),
        }
    }
}

/// A reference's strip on the desk.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReferenceStrip {
    /// The reference the strip belongs to.
    pub reference: ReferenceId,
    /// Its level, pan, mute and solo.
    pub strip: MixerStrip,
}

/// The monitor desk.
///
/// A clip or reference with no entry reads as the strips it starts with, so a source is on the
/// desk from the moment it is imported without an edit to put it there.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MixerSettings {
    /// One track per vocal clip.
    #[serde(default)]
    pub clips: Vec<ClipStrips>,
    /// One strip per reference.
    #[serde(default)]
    pub references: Vec<ReferenceStrip>,
    /// The metronome.
    pub click: MixerStrip,
}

impl Default for MixerSettings {
    fn default() -> Self {
        Self {
            clips: Vec::new(),
            references: Vec::new(),
            click: MixerStrip::new(DEFAULT_CLICK_DB, false),
        }
    }
}

impl MixerSettings {
    /// A clip's track, or the one it starts with when the desk has no entry for it.
    pub fn clip(&self, clip: ClipId) -> ClipStrips {
        self.clips
            .iter()
            .find(|entry| entry.clip == clip)
            .copied()
            .unwrap_or_else(|| ClipStrips::new(clip))
    }

    /// A reference's strip, or the one it starts with when the desk has no entry for it.
    pub fn reference(&self, reference: ReferenceId) -> MixerStrip {
        self.references
            .iter()
            .find(|entry| entry.reference == reference)
            .map_or_else(MixerStrip::default, |entry| entry.strip)
    }

    /// Every strip the desk has an entry for, the click last.
    pub fn strips(&self) -> Vec<&MixerStrip> {
        let mut strips: Vec<&MixerStrip> = Vec::new();
        for entry in &self.clips {
            strips.push(&entry.processed);
            strips.push(&entry.original);
        }
        for entry in &self.references {
            strips.push(&entry.strip);
        }
        strips.push(&self.click);
        strips
    }

    /// Whether any strip is soloed, which is what silences the ones that are not.
    pub fn soloed(&self) -> bool {
        self.strips().iter().any(|strip| strip.solo)
    }

    /// The same settings with every figure brought inside its bounds.
    ///
    /// A level out of range is clamped rather than refused, because a fader is dragged to its
    /// end rather than typed; a figure that is not a number is refused, because it is a bug. A
    /// source listed twice keeps its first entry.
    pub fn validated(&self) -> Result<Self> {
        if self.clips.len() > MAX_CLIPS || self.references.len() > MAX_REFERENCES {
            return Err(AxysError::Invalid("the desk has too many strips".into()));
        }
        let mut clips: Vec<ClipStrips> = Vec::with_capacity(self.clips.len());
        for entry in &self.clips {
            if clips.iter().any(|kept| kept.clip == entry.clip) {
                continue;
            }
            clips.push(ClipStrips {
                clip: entry.clip,
                processed: entry.processed.validated()?,
                original: entry.original.validated()?,
            });
        }
        let mut references: Vec<ReferenceStrip> = Vec::with_capacity(self.references.len());
        for entry in &self.references {
            if references
                .iter()
                .any(|kept| kept.reference == entry.reference)
            {
                continue;
            }
            references.push(ReferenceStrip {
                reference: entry.reference,
                strip: entry.strip.validated()?,
            });
        }
        Ok(Self {
            clips,
            references,
            click: self.click.validated()?,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_clip_starts_on_the_edited_take_with_the_click_below_it() {
        let mixer = MixerSettings::default();
        let track = mixer.clip(ClipId(0));
        assert!(!track.processed.mute);
        assert!(track.original.mute);
        assert!(!mixer.soloed());
        assert!(mixer.click.amplitude() < track.processed.amplitude());
    }

    #[test]
    fn a_reference_starts_at_unity() {
        let mixer = MixerSettings::default();
        assert_eq!(mixer.reference(ReferenceId(4)), MixerStrip::default());
    }

    #[test]
    fn levels_and_pans_are_clamped_rather_than_refused() {
        let mut mixer = MixerSettings::default();
        let mut track = ClipStrips::new(ClipId(0));
        track.processed.gain_db = 1000.0;
        track.original.pan = -4.0;
        mixer.clips.push(track);
        let checked = mixer.validated().expect("clamped");
        assert_eq!(
            checked.clip(ClipId(0)).processed.gain_db,
            limits::MAX_GAIN_DB
        );
        assert_eq!(checked.clip(ClipId(0)).original.pan, -1.0);
    }

    #[test]
    fn a_setting_that_is_not_a_number_is_refused() {
        let mut mixer = MixerSettings::default();
        mixer.click.gain_db = f64::NAN;
        assert!(mixer.validated().is_err());
    }

    #[test]
    fn a_source_listed_twice_keeps_its_first_entry() {
        let mut mixer = MixerSettings::default();
        let mut first = ClipStrips::new(ClipId(1));
        first.processed.gain_db = -6.0;
        mixer.clips.push(first);
        mixer.clips.push(ClipStrips::new(ClipId(1)));
        let checked = mixer.validated().expect("valid");
        assert_eq!(checked.clips.len(), 1);
        assert_eq!(checked.clip(ClipId(1)).processed.gain_db, -6.0);
    }

    #[test]
    fn a_soloed_reference_counts_as_a_solo() {
        let mut mixer = MixerSettings::default();
        mixer.references.push(ReferenceStrip {
            reference: ReferenceId(0),
            strip: MixerStrip {
                solo: true,
                ..MixerStrip::default()
            },
        });
        assert!(mixer.soloed());
    }
}
